cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAutoRebalancingReconciler.java

```diff
@@ -0,0 +1,530 @@
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static io.strimzi.api.ResourceAnnotations.ANNO_STRIMZI_IO_REBALANCE;
+import static io.strimzi.api.ResourceAnnotations.ANNO_STRIMZI_IO_REBALANCE_AUTOAPPROVAL;
+
+/**
+ * This class runs the reconciliation for the auto-rebalancing process when the Kafka cluster is scaled up/down.
+ */
+public class KafkaAutoRebalancingReconciler {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaAutoRebalancingReconciler.class.getName());
+
+    private static final String STRIMZI_IO_AUTO_REBALANCING_FINALIZER = "strimzi.io/auto-rebalancing";
+
+    private static final ScalingNodes EMPTY_SCALING_NODES = new ScalingNodes(Set.of(), Set.of());
+
+    private final Reconciliation reconciliation;
+    private final Kafka kafkaCr;
+    private final CrdOperator<KubernetesClient, KafkaRebalance, KafkaRebalanceList> kafkaRebalanceOperator;
+    private final Set<Integer> scalingDownBlockedNodes;
+    private final Set<Integer> scalingUpAddedNodes;
+
+    /**
+     * Constructs the Kafka auto-rebalancing reconciler
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param kafkaCr   The Kafka custom resource
+     * @param supplier  Supplies the operators for different resources
+     * @param scalingDownBlockedNodes  nodes blocked on scaling down because of the need for auto-rebalancing first
+     */
+    public KafkaAutoRebalancingReconciler(
+            Reconciliation reconciliation,
+            Kafka kafkaCr,
+            ResourceOperatorSupplier supplier,
+            Set<Integer> scalingDownBlockedNodes) {
+        this.reconciliation = reconciliation;
+        this.kafkaCr = kafkaCr;
+        this.kafkaRebalanceOperator = supplier.kafkaRebalanceOperator;
+        this.scalingDownBlockedNodes = scalingDownBlockedNodes;
+        this.scalingUpAddedNodes = new LinkedHashSet<>();
+    }
+
+    /**
+     * The main reconciliation method which triggers the whole reconciliation pipeline. This is the method which is
+     * expected to be called from the outside to trigger the reconciliation.
+     *
+     * @param kafkaStatus The Kafka Status class for updating auto-rebalancing status on it during the reconciliation
+     *
+     * @return  Future which completes when the reconciliation completes
+     */
+    public Future<Void> reconcile(KafkaStatus kafkaStatus) {
+        // Cruise Control is not defined in the Kafka custom resource or no autorebalance enabled, so nothing to reconcile
+        if (kafkaCr.getSpec().getCruiseControl() == null || kafkaCr.getSpec().getCruiseControl().getAutoRebalance() == null) {
+            LOGGER.infoCr(reconciliation, "Cruise Control or inner autorebalance field not defined in the Kafka custom resource, no auto-rebalancing to reconcile");
+            return Future.succeededFuture();
```

Comment:
Wdym? I can't see anything related to this kind of check in the `KafkaAssemblyOperator`. I mean, it creates a `CruiseControlReconciler` which (in its ctor) load the Cruise Control stuff from the CDR (`this.cruiseControl = CruiseControl.fromCrd`). The its reconciliation always runs but skip doing stuff if `cruiseControl` field is null.
So the CC reconciliation is always done. If CC is declared, the code here is checking that `autoRebalance` is defined otherwise the reconciliation can be skipped as well. Not sure what you have in mind.

Reply:
What I mean is whether it would make sense to check this already in KafkaAssemblyOPerator and not call this at all when this is not enabled.